lib/spack/spack/cmd/logs.py

```diff
@@ -0,0 +1,85 @@
+##############################################################################
+# Copyright (c) 2013-2017, Lawrence Livermore National Security, LLC.
+# Produced at the Lawrence Livermore National Laboratory.
+#
+# This file is part of Spack.
+# Created by Todd Gamblin, tgamblin@llnl.gov, All rights reserved.
+# LLNL-CODE-647188
+#
+# For details, see https://github.com/spack/spack
+# Please also see the NOTICE and LICENSE files for our notice and the LGPL.
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU Lesser General Public License (as
+# published by the Free Software Foundation) version 2.1, February 1999.
+#
+# This program is distributed in the hope that it will be useful, but
+# WITHOUT ANY WARRANTY; without even the IMPLIED WARRANTY OF
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the terms and
+# conditions of the GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public
+# License along with this program; if not, write to the Free Software
+# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
+##############################################################################
+
+import os.path
+
+import llnl.util.tty as tty
+import spack
+import spack.cmd
+import spack.cmd.common.arguments as arguments
+
+
+description = "fetch the logs of an installed spec"
+section = "basic"
+level = "short"
+
+
+def setup_parser(subparser):
+    subparser.add_argument(
+        '--environment',
+        action='store_true',
+        help='show the environment variables set at build-time'
+    )
+    arguments.add_common_arguments(subparser, ['constraint'])
+
+
+def logs(parser, args):
+
+    # Retrieve the specs satisfying the constraint
+    q_args = {
+        'installed': True,
+        'explicit': any,
+        'known': any
+    }
+    specs = args.specs(**q_args)
+
+    if not specs:
+        # Return early if there's no spec
+        msg = 'no spec matches your query, while exactly one is needed [{0}]'
+        msg = msg.format(' '.join(args.constraint))
+        tty.die(msg)
+
+    elif len(specs) > 1:
+        # Or if there's more than one
+        msg = 'more than one spec matches your query,'
+        msg += ' while only one is allowed [{0}]'
+        tty.error(msg.format(' '.join(args.constraint)))
+        spack.cmd.display_specs(specs, args)
+        raise SystemExit(1)
+
+    spec = specs[0]
+    if spec.external:
+        msg = '{0} is an external. External specs have no logs.'
+        tty.die(msg.format_map(spec.short_spec))
+
+    # Construct the filename to be opened
+    filename = 'build.env' if args.environment else 'build.out'
+    abs_filename = os.path.join(spec.prefix, '.spack', filename)
+
+    if not os.path.exists(abs_filename):
+        msg = 'log file does not exist! [{0}]'
+        tty.die(msg.format(abs_filename))
+
+    spack.editor(abs_filename)
```

Comment:
I think if this is just going to edit the file, it should be implemented as an argument to `spack edit`.  e.g., `spack edit --build-log <spec>`.

Similarly, if it is going to print the filename, there should probably be an argument to `spack location`.

What we don't have currently is something to just dump the log, so maybe this command should just do that?  Seems like it should also be called `spack show-log` if it does that.  But see my comments about about overlap with `spack env` for the environment log.

Reply:
`spack.editor` was the fastest way I could think of to open the logs into something paged. 

I would really like to be able to somehow open the file read only, but couldn't find a decent way to do it that works for every editor I am aware of. An alternative, to avoid accidental corruptions of the logs, is to copy them in some `tmp` space and open the copy + delete it on editor exit.

Note that if you want just a dump, for instance within a script, you can:
```
export EDITOR=cat 
spack logs <spec>
```
Does this seem reasonable to you?

Reply:
I guess it seems kind of awkward?  What about just:

```
spack show-log <spec>
```